<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Asynchronous JavaScript</title>
</head>
<body>
    <h1>Asynchronous JavaScript</h1>
    <script>
        
        // *** One ***
        /*
        const asyncFunction = () => {
            setTimeout(() => {
                console.log('...Hello, I\'m an async function, late to the party ðŸ˜³');
            }, 3000); //milliseconds
        }

        const regularFunction = () => {
            console.log('Hello - I\'m a regular fuction!');
            asyncFunction();
            console.log('The end');
        }

        regularFunction();
        */
        
        /* Two: Callbacks */
        // Simulating fetching some data from a server for recipe information using setTimeout
        /* Multiple callbacks are used within one another to simulate the data fetching, and it can become 
        messy, linking to the term 'callback hell'.
        Callback functions are performed when passed into another function, which has some job to do. It is once this job is complete that
        the callback will actually run. For example, with setTimeout, the job is to wait however long we specify. We can then run the
        callback function which we passed in, using the result of the function in the callback function. */
        
        function getRecipe()
        {
            setTimeout( () => {
                const recipeID = [45, 65, 24, 25, 69];
                console.log(recipeID);

                setTimeout(id => {
                    const recipe = {
                        title : 'Pasta',
                        publisher: 'Yusuf'};

                        console.log(`${id} : ${recipe.title}`);

                        setTimeout(publisher => {
                            const recipe2 = {
                                title: 'Pizza',
                                publisher: 'Yusuf'};
                                console.log(recipe2);
                            }, 1500, recipe.publisher);

                }, 1500, recipeID[2]);
            }, 1500);
        }
        getRecipe();
        
        /* Three: Promises
        Promises allow for you to define some sort of action to be executed, based on whether a certain event has
        taken place or not. You're able to anticipate some 'future value', defining the logic to handle such value once it arrives.
        This method of Asynchronous JavaScript eliminates the problem with callback hell, newly introduced in ES6.
        Promise States: Defines the outcome of a promise, and can be used to handle such outcomes.
        Pending - before the event happens
        Settled/Resolved - after the event has happened
         -> Result available: Promise is Fulfilled
         -> Result not available: Promise is Rejected
        */

        /* The Structure of a Promise 
        The new keyword is used to create a new instance of the Promise object, passing in a function with the arguments
        resolve and reject. We call this function the Executor Function.
        The way that the Promise has been designed allows for this executor code to run instantly, and within this function,
        we can call the resolve and reject functions which will either 'resolve' the promise - meaning a value has been retrieved
        by the executor function which we can return, or 'reject' the promise - in the event of some kind of error.
        */

        const example = new Promise( (resolve, reject) =>
        {
            // Executor Function code
        });

        /* Promise Definition */
        // Creating a promise with an executor function, defining what data should be returned upon resolved function (timer finishes)
        
        
        const getIDs = new Promise((resolve, reject) => {
            setTimeout( () => {
                resolve([45, 65, 24, 25, 69]);
            }, 1500);
        });

        // Creating a function which returns a resolved Promise object, with the recipe we want.
        const getRecipe = recID => {
            return new Promise((resolve, reject) => {
                setTimeout(ID => {
                    const recipe = {title : 'Pasta', publisher: 'Yusuf'};
                    resolve(`${ID} : ${recipe.title}`);
                }, 1500, recID);
            });
        };

        const getRelated = publisher => {
            return new Promise((resolve, reject) => {
                setTimeout(publisher => {
                    const relatedRecipe = {title: 'Pizza', publisher: 'Yusuf'};
                    resolve(`${publisher} : ${relatedRecipe.title}`);
                }, 1500, publisher);
            });
        };

        /* Promise Consumption (no async/await) */
        // Define what to do when the promise is resolved using a callback function

        getIDs
        .then(IDs => {
            // Print out the IDs, then return a *New Promise*
            console.log(IDs);
            return getRecipe(IDs[2]);
        });

        // Chaining the next then() method to the returned Promise from getRecipe
        // .then() - resolved promises, .catch() - rejected promises
        .then(recipe => {
            console.log(recipe);
            return getRelated('Yusuf');
        })
        .then(relatedRecipe => {
            console.log(relatedRecipe);
        })
        .catch(error => {
            console.log('Error');
        });
        
        /* Promise Consumption: async/await - designed to help in the consumption of Promises */

        async function getRecipesAW()
        {
            // Consuming the getIDs Promise
            const IDs = await getIDs;
            console.log(IDs);

            const recipe = await getRecipe(IDs[2]);
            console.log(recipe);

            const related = await getRelated('Yusuf');
            console.log(related);
        }

        // getRecipesAW();

        /* Async functions always return a Promise. We can use this fact to do something with the result of 
        a Promise. I have written 'newGetRecipesAW' to demonstrate this. */
        async function newGetRecipesAW()
        {
            // Consuming the getIDs Promise
            const IDs = await getIDs;
            console.log(IDs);
            const recipe = await getRecipe(IDs[2]);
            console.log(recipe);
            return recipe;
        }

        // Using the then() method on the returned promise, simply console logging the result.
        newGetRecipesAW().then(result => console.log(`${result} is the best ever!`));

        // AJAX Calls with Fetch and Promise
        const getWeatherInfo = woeid =>
        {
            // fetch(): Returns a Promise to be consumed 
            fetch(`https://www.metaweather.com/api/location/${woeid}`)
            .then(result => {
                console.log(result);

                // JavaScript Object -> JSON
                // Promise is returned from the json()
                return result.json();
            })
            .then(data => {
                // Retrieving data from JSON and displaying to the console
                const today = data.consolidated_weather[0];
                console.log(`Temperatures in ${data.title} stay between ${today.min_temp} and ${today.max_temp}`);
            })
            .catch(error => console.log(error));
        };

        // getWeatherInfo(28218);

        // AJAX Calls with Fetch and Async / Await
        // Reminder: Async functions always return a Promise
        const getWeatherAW = async woeid => {
            try
            {
                const result = await fetch(`https://www.metaweather.com/api/location/${woeid}`);
                const data = await result.json();
                //console.log(data);
                const tomorrow = data.consolidated_weather[1];
                console.log(`Temperatures in ${data.title} stay between ${tomorrow.min_temp} and ${tomorrow.max_temp}`);

                return data;
            } catch(error) {
                console.log(error);
            }
        }

        // Executing getWeather for Manchester
        getWeatherAW(28218);

        let dataLondon = '';

        // Using the returned Promise from getWeatherAW to store the data in dataLondon
        getWeatherAW(44418).then(data => {
            dataLondon = data;
            console.log(dataLondon);
        });
        

        // MDN Promises Exercise

        // fetch() returns a Promise which we store in the variable 'promise'
        // When the Promise is neither a success or failure, it is in a Pending state
        let promise = fetch('coffee.jpg');

        // Invoking the .then() method lets us perform some kind of action upon a successful Promise,
        // upon which a Response object is returned - also understood to be when a promise has been 'fulfilled'.
        // Below, upon our first Promise completing, we run the .blob() method which transform the image into a 
        // Blob object to allow for some operations to be done on the image.
        // The .then() method itself will return a Promise. 
        let promise2 = promise.then(response => response.blob());

        let promise3 = promise2.then(myBlob => {
            let objectURL = URL.createObjectURL(myBlob);
            let image = document.createElement('img');
            image.src = objectURL;
            document.body.appendChild(image);
        });

        // Invoking the .catch() method allows for us to handle a failure in a Promise
        let errorCase = promise3.catch(e => console.log(`Problem with fetch operation: ${e.message}`));
        
    </script>
</body>
</html>